### DISCUSSION TITLE
Remote profiling with async profiler

### DISCUSSION THREAD

**Discussion by kibor** [async-profiler/async-profiler#496] [Q&A] [answered]
Good day everyone,

Sorry if this question had been discussed already, I didn't find appropriate topics. Actually, this is 2 questions that are tightly coupled with each other.

1. Is it possible to control async profiler remotely from another computer?
I have seen code related to JMX in code source, so I guess it somewhat answers my question, but JMX seems to be purely for Java applications. When client that sends control commands is not written in Java (Golang, for example), it seems to be a problem.

BTW, if JMX is actually a viable approach here, can you please add an example file (like profile.sh) to the repository, just to show how to use it?

2. Bundling application with async profiler
To me it seems to be much easier, when I don't have to make sure that proper so file is located at the correct path on any server that app is being deployed on. Ideally app dependencies should be bundled with the app itself.
Is it possible to add some kind of Maven dependency to POM file, so result jar already contains so file needed for async profiler to work?

**Comment by toktarev** [comment]
--Bundling application with async profiler

Yes it is possible to run profiler in embedded mode.
you should use java api, see one.profiler.AsyncProfiler

---Is it possible to add some kind of Maven dependency to POM file, so result jar already contains so file needed for async profiler to work?

I've added ".so" file into resources folder and it automatically appeared in the application jar file.

**Reply by kibor** [reply]
> I've added ".so" file into resources folder and it automatically appeared in the application jar file.

Right, I can easily do it for my own app, but it can be a problem, when there are multiple services, and they all need to do it for themselves.
One more point is version control - with POM file it is much easier.
Ideally, each service just adds one extra dependency to its POM file, and afer final jar is ready, it contains ".so" file that is needed for async profiler. Is that possible?

**Reply by krzysztofslusarski** [reply]
There is already an issue for that purpose #93. It is not done yet.

**Reply by kibor** [reply]
Sorry, missed that one.

**Comment by apangin** [comment, marked as answer]
Async-profiler can be run in many ways: as an agent, from an external process using dynamic attach, within a Java application using Java API, or via JMX. It is a low level tool, and there is deliberately no built-in mechanism to control it remotely, since remote control is a job of a higher layer. I don't think the profiler library itself should handle things like REST API, web UI, authentication, etc.

So you may take, for instance, the profiler Java API and drop it into a web framework of your preference - it's just a matter of few lines of code.

Or, as you already noticed, async-profiler can integrate with JMX. One possible option to do it is to register AsyncProfiler MXBean:
```
ManagementFactory.getPlatformMBeanServer().registerMBean(
        new StandardMBean(AsyncProfiler.getInstance(), null, true),
        new ObjectName("one.profiler:type=AsyncProfiler"));
```

After that, you'll find `AsyncProfiler` MXBean in jconsole / VisualVM / Mission Control / etc, and will be able to call its methods.

Another option to invoke async-profiler via JMX is to use built-in DiagnosticCommand MBean: `com.sun.management:type=DiagnosticCommand`. It does not require registering AsyncProfiler, but works only since JDK 11.
Find `jvmtiAgentLoad` operation and invoke it with two arguments: the path to `libasyncProfiler.so` and the [profiler options string](https://github.com/jvm-profiling-tools/async-profiler/blob/master/src/arguments.cpp#L51).

**Reply by andrewazores** [reply]
@apangin sorry to resurrect such an old topic, but this seemed like a relevant place to ask.

I'm looking at integrating some support for async-profiler [into Cryostat](https://github.com/cryostatio/cryostat/issues/643). The AsyncProfilerMXBean looks like it would be a good integration point, but from what I can tell [this requires application code](https://github.com/async-profiler/async-profiler/blob/fde780e2759606561f2d30cabe7b0def70d392af/src/api/one/profiler/AsyncProfilerMXBean.java#L12) to actually register into the MBean server. I *could* (and have prototyped) something for the cryostat-agent to chain load async-profiler, but there are a few reasons I prefer not to go that route [0].

[This other prior discussion](https://github.com/async-profiler/async-profiler/issues/125) also seems relevant - would you be open to an idea like using `-javaagent` and associated entry points (ie. `premain`/`agentmain`) as a way to bootstrap into `AsyncProfiler.getInstance()` and registering the MXBean? Or would you have some other preferred method in mind for how to get the AsyncProfilerMXBean registered in the user's application via some config rather than adding supporting code? If you're open to this idea I would be open to proposing a PR :-)

[0] this might tie async-profiler versioning into Cryostat versioning and I think it's best for the user to be able to control that. It also would only do anything for Cryostat users who are using the Cryostat Agent too, and not ones who go without the Agent and simply use Cryostat via JMX. And of course, the MXBean interface feels like an appropriately loose coupling between components.

**Reply by apangin** [reply]
@andrewazores Thank you for your interest in async-profiler. Can you expand more on the essense of the problem? I believe Cryostat can already start and stop JDK Flight Recorder dynamically. How is async-profiler different? Why can't you do the same?

**Reply by andrewazores** [reply]
@apangin thanks for the quick response. I have already prototyped out some integration between Cryostat and async-profiler where I can get Cryostat to start, stop, and retrieve data from async-profiler dynamically, but in order to do so I added async-profiler as a Maven dependency in the cryostat-agent. cryostat-agent gets attached to the target application via `-javaagent`, and in my prototype I have cryostat-agent calling into AsyncProfiler.getInstance() and using the async-profiler Java API. So from a functionality perspective, this works.

But from the distribution perspective, bundling async-profiler as a direct dependency within the cryostat-agent shaded JAR is not ideal. I'd like to be able to just query the target application's MBean server over JMX and check if the AsyncProfilerMXBean is present, just like how we do for the FlightRecorderMXBean, and perform all remote management over that interface. Then the user does not need to use cryostat-agent either, and it's up to the user to add async-profiler if they want to use this integration. So here is the root of the (perceived) problem - if I decouple cryostat-agent from async-profiler, I don't think there is any more way for the AsyncProfilerMXBean to become present in the target application, unless I tell the end user that they have to add code into their application to do that setup and register the MXBean themselves. If I could instead tell end users that rather than adding code, recompiling, redeploying, etc. their applications they can simply add a `-javaagent` flag or add some other configuration instead, then I think that would be a better experience. It would at least match more closely with the user story we tell that Cryostat can be used without having to add any code or do any application rebuilds. The application can be the same and just some configuration or deployment properties change, ex. mounting some volume and editing some environment variable.

**Reply by apangin** [reply]
@andrewazores I see what you mean now, thanks. This should be fairly easy to implement. Please submit a new [Feature Request](https://github.com/async-profiler/async-profiler/issues/new?template=feature-request.yml), and we'll consider it for one of the next releases.

**Reply by andrewazores** [reply]
#1627
